### PR TITLE
Implement formdata event

### LINKED ISSUE #22654: Implement formdata event
https://github.com/whatwg/html/pull/4239
https://github.com/whatwg/xhr/pull/231
https://github.com/web-platform-tests/wpt/pull/14637

This specification change adds an event that is fired while submitting a form, allowing web content to modify the form contents before it is submitted.

### PR DESCRIPTION
**Pull request by CYBAI**
---
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #22654 
- [x] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/22660)

### CODE REVIEW

File: tests/wpt/web-platform-tests/xhr/formdata.htm
Comment (CYBAI) [inline tests/wpt/web-platform-tests/xhr/formdata.htm]: I cannot pass the original test case locally and then I found it might be more reasonable to test with `formDataInEvent` here? because it's assigned with the appended `e.formData` which should have `h1`.

File: components/script/dom/htmlformelement.rs
Comment (jdm) [inline components/script/dom/htmlformelement.rs]: Could you implement this check? The algorithm at https://html.spec.whatwg.org/multipage/links.html#cannot-navigate is straightforward, and there are three uses of it in the form submissions algorithm.

File: components/script/dom/formdataevent.rs
Comment (jdm) [inline components/script/dom/formdataevent.rs]: Let's use `formData` in this string.

File: components/script/dom/formdataevent.rs
Comment (jdm) [inline components/script/dom/formdataevent.rs]: No need for this function; there are no callers so we can inline it in `new` instead.

File: components/script/dom/formdata.rs
Comment (jdm) [inline components/script/dom/formdata.rs]: We can remove this TODO. Also let's add a link to https://xhr.spec.whatwg.org/#dom-formdata.

File: components/script/dom/webidls/FormDataEvent.webidl
Comment (jdm) [inline components/script/dom/webidls/FormDataEvent.webidl]: What happens if we uncomment this? If there's an error, do we have an issue filed for it, and can we add a comment?

File: tests/wpt/web-platform-tests/xhr/formdata.htm
Comment (jdm) [reply tests/wpt/web-platform-tests/xhr/formdata.htm]: This change should not be necessary if my earlier comment is addressed.

**Review by jdm — CHANGES_REQUESTED** [review]
Great work! These changes were easy to read and understand.

File: components/script/dom/htmlformelement.rs
Comment (CYBAI) [reply components/script/dom/htmlformelement.rs]: Sure! Let me try to implement it!

File: components/script/dom/formdata.rs
Comment (CYBAI) [reply components/script/dom/formdata.rs]: Thanks for checking this one! I was confused if I should remove it or not!

File: components/script/dom/webidls/FormDataEvent.webidl
Comment (CYBAI) [reply components/script/dom/webidls/FormDataEvent.webidl]: It will fail with this compilation error. I didn't find a corresponding issue. I'll try to read the webidl codegen to see what the root cause is later.

```rust
error[E0599]: no function or associated item named `empty` found for type `dom::bindings::codegen::Bindings::FormDataEventBinding::FormDataEventInit` in the current scope
   --> /Users/cybai/codespace/mozilla/servo/target/debug/build/script-59d4418593395066/out/Bindings/FormDataEventBinding.rs:857:95
    |
259 | pub struct FormDataEventInit {
    | ---------------------------- function or associated item `empty` not found for this
...
857 |             crate::dom::bindings::codegen::Bindings::FormDataEventBinding::FormDataEventInit::empty()
    |             ----------------------------------------------------------------------------------^^^^^
    |             |
    |             function or associated item not found in `dom::bindings::codegen::Bindings::FormDataEventBinding::FormDataEventInit`
```

File: components/script/dom/htmlformelement.rs
Comment (CYBAI) [reply components/script/dom/htmlformelement.rs]: I can udnerstand why the testcase failed now but I'm still trying to figure out how to avoid the `clone` here 🤔 Would it be a right way to pass the first argument of `FormData::new` with `Option<&Vec<FormDatum>>`?

**Comment by CYBAI** [comment]
Just updated some easy fix and remove the `update test case` commit! Will keep working on addressing other comments!

File: components/script/dom/htmlformelement.rs
Comment (jdm) [reply components/script/dom/htmlformelement.rs]: I expect you want to pass `ret` itself, then add an API to FormData to extract it again later and return that value.

Comment (CYBAI) [reply components/script/dom/htmlformelement.rs]: Ah! Yeah! I was thinking from constructor perspective 😂 That must be a good solution to extract it again! Thanks!

Comment (CYBAI) [reply components/script/dom/htmlformelement.rs]: Just found there's an existing API to extract the datums.

https://github.com/servo/servo/blob/192fbafd11126744a299558d14c4b2c63e3cd2ca/components/script/dom/formdata.rs#L197-L204

One more question for this; not sure why there're two `impl FormData` but maybe it's better to combine them? 🤔

Comment (jdm) [reply components/script/dom/htmlformelement.rs]: Combining them is fine.

File: components/script/dom/htmlformelement.rs
Comment (CYBAI) [reply components/script/dom/htmlformelement.rs]: Implemented in 038c83b

File: components/script/dom/element.rs
Comment (jdm) [inline components/script/dom/element.rs]: `!self.is::<HTMLAnchorElement>()` is easier to read :)

File: components/script/dom/element.rs
Comment (jdm) [inline components/script/dom/element.rs]: This is a method on Element, so we should be able to call `self.is_connected()`

File: components/script/dom/webidls/FormDataEvent.webidl
Comment (jdm) [reply components/script/dom/webidls/FormDataEvent.webidl]: Ok, it is clear to me why this happens now. Filed https://github.com/servo/servo/issues/22670 about it.

**Review by jdm — CHANGES_REQUESTED** [review]
Looking good!

File: components/script/dom/element.rs
Comment (CYBAI) [reply components/script/dom/element.rs]: Ah, I forgot to update it! Thanks for catching this 😂

File: components/script/dom/element.rs
Comment (CYBAI) [reply components/script/dom/element.rs]: Yeah! This is better!

File: components/script/dom/webidls/FormDataEvent.webidl
Comment (CYBAI) [reply components/script/dom/webidls/FormDataEvent.webidl]: Thanks a lot! After reading the issue, I can understand why it failed to compile with `required` :)

File: tests/wpt/metadata/html/semantics/forms/form-submission-0/constructing-form-data-set.html.ini
Comment (CYBAI) [inline tests/wpt/metadata/html/semantics/forms/form-submission-0/constructing-form-data-set.html.ini]: This test case still gets failure.

https://github.com/servo/servo/blob/4ac5b8a3aea5df3bf374dcfdb22c34f5ff0bafe8/tests/wpt/web-platform-tests/html/semantics/forms/form-submission-0/constructing-form-data-set.html#L98-L109

I tried to add logs to check if the event is dispatched and it's triggered correctly.
So I'm wondering if the `search` is not updated correctly after submitting the form.

Comment (CYBAI) [reply tests/wpt/metadata/html/semantics/forms/form-submission-0/constructing-form-data-set.html.ini]: Ok. I've found what exact the wrong output is, will keep investigating the root cause.

The expected search is `n1=v1&h1=vh1` but we got `h1=vh1&n1=v1`.

Comment (CYBAI) [reply tests/wpt/metadata/html/semantics/forms/form-submission-0/constructing-form-data-set.html.ini]: I've observed one issue here but the test case still got failure.

About the observed issue is, we're using `BTreeMap` for `FormData` so when we do iteration for it, the data will be sorted automatically. I'm wondering maybe we need to change to `IndexMap` instead?

https://github.com/servo/servo/blob/4d94692b043479ef420c51eea2b3fc9e7e709232/components/script/dom/formdata.rs#L28

After observed above issue, I tried to use `IndexMap` but I found the test case still got failure. Then I tried to make a break point in this line

https://github.com/servo/servo/blob/4d94692b043479ef420c51eea2b3fc9e7e709232/components/script/dom/htmlformelement.rs#L471

```
(lldb) p *load_data.url.__0.ptr.pointer
(alloc::sync::ArcInner<url::Url>) $1 = {
  strong = {
    v = (value = 1)
  }
  weak = {
    v = (value = 1)
  }
  data = {
    serialization = {
      vec = {
        buf = {
          ptr = (pointer = "http://web-platform.test:8000/common/blank.html?n1=v1&h1=vh1", _marker = core::marker::PhantomData<unsigned char> @ 0x00007fbc259b51f0)
          cap = 94
        }
        len = 60
      }
    }
    scheme_end = 4
    username_end = 7
    host_start = 7
    host_end = 24
    host = {}
    port = {}
    path_start = 29
    query_start = {}
    fragment_start = {}
  }
}
```

I found the url was shown correctly with `?n1=v1&h1=vh1`.

But, when I added a log `console.log(iframe.contentWindow.location.href);` in the `iframe.onload` event listener. It shows the `href` is `about:blank` so the `location.search` would be empty string which will not be equal to `n1=v1&h1=vh1`.

I'm wondering if the onload event is triggered after the form submitted 🤔

Comment (jdm) [reply tests/wpt/metadata/html/semantics/forms/form-submission-0/constructing-form-data-set.html.ini]: Ok, I did some investigation around iframe load events (which caused me to read through https://github.com/whatwg/html/issues/490 and remember a bunch of awful details). As far as I can tell, the reason we observe the about:blank behaviour in this test is that there is a load event queued for the iframe immediately when it is added to the document (for the default about:blank browsing context). I created a test to verify this theory:
```html
<script src="/resources/testharness.js"></script>
<script src="/resources/testharnessreport.js"></script>
<iframe name="i" id="i"></iframe>
<form id="f" action="/common/blank.html" target="i"></form>
<script>
  async_test(function(t) {
    var i = document.getElementById('i');
    i.onload = t.step_func(function() {
      if (i.contentWindow.location.href == "about:blank") { return; } 
      assert_equals(i.contentWindow.location.pathname, "/common/blank.html"); 
      t.done();
    });
    document.getElementById('f').submit();
  });
</script>
```
This passes in Servo for me today. The spec says that the iframe load event should be queued when it's added to the document, and that form submission causes a task to be queued (and that task initiates a navigation for the iframe's browsing context), so it seems perfectly valid for the about:blank's load event to fire during that time before the iframe has officially been navigated away from about:blank.

THEREFORE

I think the right solution here is to add a comment with a link to https://github.com/whatwg/html/issues/490 and add a check for about:blank in the iframe load event handler and ignore the event if it matches.

**Review by jdm — APPROVED** [review]
This looks good apart from the change to the test that I suggested!

File: tests/wpt/metadata/html/semantics/forms/form-submission-0/constructing-form-data-set.html.ini
Comment (CYBAI) [reply tests/wpt/metadata/html/semantics/forms/form-submission-0/constructing-form-data-set.html.ini]: @jdm thanks a lot for helping me investigate this! That's clear for me now that why the test failed; I think we can only update the test and remove the comment once the issue is fixed.

File: tests/wpt/web-platform-tests/html/semantics/forms/form-submission-0/constructing-form-data-set.html
Comment (jdm) [inline tests/wpt/web-platform-tests/html/semantics/forms/form-submission-0/constructing-form-data-set.html]: This will cause the test to finish after the about:blank load event. This should be `step_func` and there should be a `t1.done()` after the assert instead.

**Review by jdm — CHANGES_REQUESTED** [review]

File: tests/wpt/web-platform-tests/html/semantics/forms/form-submission-0/constructing-form-data-set.html
Comment (jdm) [inline tests/wpt/web-platform-tests/html/semantics/forms/form-submission-0/constructing-form-data-set.html]: I think a more clear comment would be:
```
// The initial about:blank load event can be fired before the form navigation occurs.
// See https://github.com/whatwg/html/issues/490 for more information.
```

File: tests/wpt/web-platform-tests/html/semantics/forms/form-submission-0/constructing-form-data-set.html
Comment (CYBAI) [reply tests/wpt/web-platform-tests/html/semantics/forms/form-submission-0/constructing-form-data-set.html]: Ahhh, yes! Thanks for the good catch! I only check it passed 😓 Let me update!

File: tests/wpt/web-platform-tests/html/semantics/forms/form-submission-0/constructing-form-data-set.html
Comment (CYBAI) [reply tests/wpt/web-platform-tests/html/semantics/forms/form-submission-0/constructing-form-data-set.html]: Thanks! This is indeed much more clear!

**Comment by jdm** [comment]
@bors-servo delegate+
This needs `./mach update-manifest`, then it can merge.

**Comment by CYBAI** [comment]
@jdm Thanks! Waiting #22736 merged! I'll rebase and `update-manifest`!

**Comment by jdm** [comment]
Looks like the webhook just missed the previous notification, since this PR did not show up in the homu queue.

**Comment by CYBAI** [comment]
Maybe I should wait a minute to ask homu to approve after I force pushed 😂 Thanks for the approval!

**Pull request merged**
